apps/passport-client/src/dispatch.ts

```diff
@@ -942,3 +956,67 @@ async function promptToAgreePrivacyNotice(state: AppState, update: ZuUpdate) {
     });
   }
 }
+
+/**
+ * Merge in a PCD collection from an import of backed-up data.
+ */
+async function mergeImport(
+  state: AppState,
+  update: ZuUpdate,
+  collection: PCDCollection,
+  pcdsToMergeIds: Set<PCD["id"]>
+) {
+  const userHasExistingSemaphoreIdentityPCD =
+    state.pcds.getPCDsByType(SemaphoreIdentityPCDTypeName).length > 0;
+
+  const userHasExistingEmailPCD =
+    state.pcds.getPCDsByType(EmailPCDTypeName).length > 0;
+
+  const pcdCountBeforeMerge = state.pcds.getAll().length;
+
+  const filterFunction = (pcd: PCD, target: PCDCollection): boolean => {
+    return (
+      pcdsToMergeIds.has(pcd.id) &&
+      !(pcd.type === EmailPCDTypeName && userHasExistingEmailPCD) &&
+      !(isSemaphoreIdentityPCD(pcd) && userHasExistingSemaphoreIdentityPCD) &&
+      !(
+        isSemaphoreIdentityPCD(pcd) &&
+        pcd.claim.identity.getCommitment().toString() !== state.self.commitment
+      ) &&
+      !target.hasPCDWithId(pcd.id)
+    );
+  };
+
+  const packages = await getPackages();
+
+  // This is a clunky way to clone an object.
+  // It is also asynchronous, so there is the infinitessimal possibility of
+  // an update changing state.pcds while we're in the middle of deserializing.
+  const pcds = await PCDCollection.deserialize(
+    packages,
+    await state.pcds.serializeCollection()
+  );
+
+  try {
+    pcds.merge(collection, {
+      filter: filterFunction
+    });
+
+    await savePCDs(pcds);
+    update({
+      pcds,
+      modal: {
+        modalType: "account-import",
```

Comment:
Is there a race here if the user closes the modal while import is in progress, meaning the modalType is being clobbered here?  Or is the point of a modal that the user can't do that?